common/src/com/thoughtworks/go/config/materials/git/GitMaterial.java

```diff
@@ -106,28 +127,32 @@ public MaterialConfig config() {
     }
 
     public List<Modification> modificationsSince(File baseDir, Revision revision, final SubprocessExecutionContext execCtx) {
+        unshallowIfNeeded(ProcessOutputStreamConsumer.inMemoryConsumer(), revision, baseDir);
         GitCommand gitCommand = getGit(baseDir);
         return gitCommand.modificationsSince(revision);
     }
 
     public MaterialInstance createMaterialInstance() {
-        return new GitMaterialInstance(url.forCommandline(), branch, submoduleFolder, UUID.randomUUID().toString());
+        return new GitMaterialInstance(url.forCommandline(), branch, submoduleFolder, UUID.randomUUID().toString(), shallowClone);
     }
 
     @Override
     protected void appendCriteria(Map<String, Object> parameters) {
         parameters.put(ScmMaterialConfig.URL, url.forCommandline());
         parameters.put("branch", branch);
+        parameters.put("shallowClone", shallowClone);
```

Comment:
Yes, fingerprint is used all over. My point is, why can't this value be like `submoduleFolder` which is in GitMaterial? It's not part of the fingerprint, but [is persisted](https://github.com/gocd/gocd/blob/eb0ae575fbb0a30f59ed646cd45d16eb57d26aa5/server/resources/hibernate-mappings/MaterialInstance.hbm.xml#L46).

Also, as a part of pluggable SCM support, a new field called "additionalData" was introduced. It can be used in any material and can hold a JSON value. Is that something we should look into for this? I ask this because: As you know, the materials table is flattened. So, even a GitMaterial row in it has a null value for a column called `useTickets`, which is a Perforce (@jyotisingh's favorite SCM) field. So, the idea of "additionalData" as a column was to have anything which is specific to that SCM, but not necessarily common to all. "shallowClone" fits that definition according to me.

Whatever we choose, I wonder if there's a reason to include shallowClone in the fingerprint. Assume it is not part of the fingerprint, and so the Go Server will have only one material for it, and won't differentiate between the two. But, the Go Server is not the goal of this PR. The agents are the goal (where a shallow or non-shallow checkout really matters).

I know it kind of matters in Go Server as well, but I don't know if this is big enough of a reason to change fingerprint. Migrating fingerprint, as you say, might be an option. We'd need to ensure that it is correct, though.